### PR TITLE
complete rewrite of `.unitConverter()` using base-R

### PR DESCRIPTION
**Pull request by IndrajeetPatil**
``` r
library(ospsuite)
#> Loading required package: rClr
#> Loading the dynamic library for Microsoft .NET runtime...
#> Loaded Common Language Runtime version 4.0.30319.42000

df <- dplyr::tibble(
  xValues = c(15, 30, 60), xUnit = "min", xDimension = "Time",
  yValues = c(0.25, 45, 78), yUnit = c("", "%", "%"), yDimension = c("Fraction", "Fraction", "Fraction"),
  molWeight = c(10, 10, 10)
)

ospsuite:::.unitConverter(df)
#> [1] "ping for  min"
#> [1] "ping for  "
#> [1] "ping for  %"
#> # A tibble: 3 x 7
#>   xValues xUnit xDimension yValues yUnit yDimension molWeight
#>     <dbl> <chr> <chr>        <dbl> <chr> <chr>          <dbl>
#> 1      15 min   Time          0.25 ""    Fraction          10
#> 2      30 min   Time          0.45 ""    Fraction          10
#> 3      60 min   Time          0.78 ""    Fraction          10
```

<sup>Created on 2022-04-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

### CODE REVIEW

File: R/unitConverter.R [hunk 1]
```diff
@@ -50,45 +50,58 @@
...
-        dplyr::ungroup() %>%
-        tidyr::unnest(cols = c(yValues)) %>%
-        dplyr::mutate(yUnit = yTargetUnit) %>%
-        dplyr::select(-data)
+  yDataList <- split(data, data$yUnit)
+  data <- dplyr::bind_rows(lapply(yDataList, .yUnitConverter, yTargetUnit))
 
-      # combine them
-      dplyr::bind_cols(dataXUnit, dataYUnit)
+  data
```
Comment (PavelBal) [inline R/unitConverter.R:63]: Please add `return(data)` to all functions that return something.

Comment (IndrajeetPatil) [reply R/unitConverter.R:63]: Yes, please don't review this until I mark it ready for a review.

Comment (PavelBal) [reply R/unitConverter.R:63]: OK. Can I push a test for different MWs (is failing now)?

**Comment by IndrajeetPatil** [comment]
Example with error units:

``` r
library(ospsuite)
#> Loading required package: rClr
#> Loading the dynamic library for Microsoft .NET runtime...
#> Loaded Common Language Runtime version 4.0.30319.42000

dfError <- dplyr::tibble(
  xValues = c(15, 30, 60), xUnit = "min", xDimension = "Time",
  yValues = c(0.25, 45, 78), yUnit = c("", "%", "%"), yDimension = c("Fraction", "Fraction", "Fraction"),
  yErrorValues = c(0.01, 5, 8), yErrorUnit = c("", "%", "%"),
  molWeight = c(10, 10, 10)
)

ospsuite:::.unitConverter(dfError)
#> [1] "ping for  min"
#> [1] "ping for  "
#> [1] "ping for  %"
#> [1] "ping for  "
#> [1] "ping for  %"
#> # A tibble: 3 x 9
#>   xValues xUnit xDimension yValues yUnit yDimension yErrorValues yErrorUnit
#>     <dbl> <chr> <chr>        <dbl> <chr> <chr>             <dbl> <chr>     
#> 1      15 min   Time          0.25 ""    Fraction           0.01 ""        
#> 2      30 min   Time          0.45 ""    Fraction           0.05 ""        
#> 3      60 min   Time          0.78 ""    Fraction           0.08 ""        
#> # ... with 1 more variable: molWeight <dbl>
```

<sup>Created on 2022-04-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

**Comment by IndrajeetPatil** [comment]
I can't commit now because of a merge conflict 🤦

image

**Comment by PavelBal** [comment]
Pull first?

**Comment by IndrajeetPatil** [comment]
> Pull first?

I had already committed, so couldn't push or pull without resolving the merge conflict first. 

I undid my last commit, pulled, and then pushed again.

**Comment by IndrajeetPatil** [comment]
With different molecular weights:

``` r
library(ospsuite)
#> Loading required package: rClr
#> Loading the dynamic library for Microsoft .NET runtime...
#> Loaded Common Language Runtime version 4.0.30319.42000

df <- dplyr::tibble(
  xValues = c(15, 30, 60), xUnit = "min", xDimension = "Time",
  yValues = c(1, 1, 1), yUnit = c("mol", "mol", "mol"), yDimension = c(ospDimensions$Amount, ospDimensions$Amount, ospDimensions$Amount),
  molWeight = c(10, 10, 20)
)

ospsuite:::.unitConverter(df, yUnit = "g")
#> [1] "ping for  min"
#> [1] "ping for  mol"
#> [1] "ping for  mol"
#> # A tibble: 3 x 7
#>   xValues xUnit xDimension yValues yUnit yDimension molWeight
#>     <dbl> <chr> <chr>        <dbl> <chr> <chr>          <dbl>
#> 1      15 min   Time            10 g     Amount            10
#> 2      30 min   Time            10 g     Amount            10
#> 3      60 min   Time            20 g     Amount            20
```

<sup>Created on 2022-04-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

**Comment by IndrajeetPatil** [comment]
This PR is ready for a review now.

**Comment by msevestre** [comment]
What does this PR fix? What is the corresponding issue?

**Comment by IndrajeetPatil** [comment]
The issue is already linked in the PR:

image

**Comment by IndrajeetPatil** [comment]
> Why is this method not exposed?

cf. https://github.com/Open-Systems-Pharmacology/OSPSuite-R/pull/835#issuecomment-1074945565

> Looks good and indeed easier to read for normal people.

*normal* 😆

Yeah, better to not write _too_ clever code only to reduce lines of code, but effectively putting that code in a silo away from the rest of the team.

Screenshot 2022-04-12 at 22 03 45

**Comment by msevestre** [comment]
Looks good to me. Merging.

**Pull request merged**